File: LeetCode/Study Plan/Practice Problems/Medium/Asteroid Collision/Go/asteroidCollision.go
package main

import "fmt"

func asteroidCollision(asteroids []int) []int {
	size := len(asteroids)
	stack := []int{}

	for i := 0; i < size; i++ {
		if len(stack) == 0 {
			stack = append(stack, asteroids[i])
			continue
		}

		if asteroids[i] < 0 && stack[len(stack)-1] > 0 {
			push := false
			for len(stack) > 0 && (asteroids[i] < 0 && stack[len(stack)-1] > 0) {
				if abs(asteroids[i]) > stack[len(stack)-1] {
					stack = stack[:len(stack)-1]
					push = true
				} else if abs(asteroids[i]) < stack[len(stack)-1] {
					push = false
					break
				} else {
					push = false
					stack = stack[:len(stack)-1]
					break
				}
			}
			if push {
				stack = append(stack, asteroids[i])
			}
		} else {
			stack = append(stack, asteroids[i])
		}
	}

	return stack
}

func abs(a int) int {
	if a > 0 {
		return a
	}
	return -1 * a
}

func main() {
	fmt.Println(asteroidCollision([]int{5, 10, -5}))
}

File: LeetCode/Study Plan/Practice Problems/Medium/Asteroid Collision/Java/asteroidCollision.java
class Solution {
    public int[] asteroidCollision(int[] asteroids) {
        int size = asteroids.length;
        Stack<Integer> s = new Stack<>();

        for (Integer aster : asteroids) {
            if (s.isEmpty()) {
                s.add(aster);
                continue;
            }
            
            if (aster<0 && s.peek().equals(Math.abs(aster))) {
                s.pop();
            }else {
                boolean push = true;
                while (!s.isEmpty() && (s.peek() > 0 && aster < 0)) {
                    if (Math.abs(aster) > s.peek()) {
                        s.pop();
                    }else if (Math.abs(aster) == s.peek()) {
                        push = false;
                        s.pop();
                        break;
                    }else {
                        push = false;
                        break;
                    }
                }
                if (push) s.add(aster);
            }
        }
        int stackSize = s.size();
        int[] res = new int[s.size()];

        for (int i=stackSize-1; i>=0; i--) {
            res[i] = s.pop();
        }

        return res;
    }
}
